src/libraries/Microsoft.Extensions.Configuration/src/InternalConfigurationRootExtensions.cs

```diff
@@ -22,13 +22,48 @@ internal static IEnumerable<IConfigurationSection> GetChildrenImplementation(thi
         {
             using ReferenceCountedProviders? reference = (root as ConfigurationManager)?.GetProvidersReference();
             IEnumerable<IConfigurationProvider> providers = reference?.Providers ?? root.Providers;
+            var optimizedProviders = new List<ConfigurationProvider>();
+            bool allProvidersAreOptimized = true;
+            foreach (IConfigurationProvider provider in providers)
+            {
+                // If a IConfigurationProvider inherit from ConfigurationProvider, but override the implementation GetChildKeys
+                // To make sure we get the expected keys, we should honor the override and go the legacy path.
+                Type type = provider.GetType();
+                if (provider is not ConfigurationProvider optimizedProvider
+                    || type.GetMethod(nameof(provider.GetChildKeys))?.DeclaringType == type)
+                {
+                    allProvidersAreOptimized = false;
```

Comment:
I'm inclined to agree @rjgotten that `GetChildKeys` is a leaky abstraction.

I know we've avoided default interface methods in Microsoft.Extensions.* because of the nestandard2.0 target, but this would be a great candidate for it. @eerhardt do you know if it's possible to make this work at least for newer targets?

If we could add a default interface method, `IConfigurationProvider.GetChildKeys(string? parentPath)` could call into the existing `GetChildKeys(IEnumerable<string> earlierKeys, string? parentPath)` by default, but the new `parentPath`-only overload would be optimized because it would not need to guarantee sorting like the existing method.

Reply:
By adding the default interface method, I run into the follow apicompat issue. Any idea?

```
6>Done building project "Microsoft.Extensions.Configuration.csproj" -- FAILED.
6>C:\Users\yuantw\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22419.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : Compat issues with assembly Microsoft.Extensions.Configuration:
6>C:\Users\yuantw\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22419.1\build\Microsoft.DotNet.ApiCompat.targets(95,5): error : MembersMustExist : Member 'public System.Collections.Generic.IEnumerable<System.String> Microsoft.Extensions.Configuration.ConfigurationProvider.GetKeys(System.String)' does not exist in the reference but it does exist in the implementation.
6>C:\Users\yuantw\.nuget\packages\microsoft.dotnet.apicompat\7.0.0-beta.22419.1\build\Microsoft.DotNet.ApiCompat.targets(112,5): error : MatchingRefApiCompat failed - The reference assembly doesn't match all the APIs in the implementation for 'D:\runtime\artifacts\bin\Microsoft.Extensions.Configuration\Debug\netstandard2.0\Microsoft.Extensions.Configuration.dll'. To address either fix errors in the reference assembly (referenced as implementation in compat errors for this reverse compat check), add the issues to the baseline file '' or disable this check by setting RunMatchingRefApiCompat=false in this project.
6>Done building project "Microsoft.Extensions.Configuration.csproj" -- FAILED.
```


I guess the problem is: if user upgrade nuget package version of `Microsoft.Extensions.Configuration`, but didn't upgrade the nuget package version of `Microsoft.Extensions.Configuration.Abstractions`, then it wil throw `MissingMethodException` in runtime. I run into the exact the problem during benchmark - at first I forgot to copy `Microsoft.Extensions.Configuration.Abstractions.dll` to `testhost` folder.

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.MissingMethodException: Method not found: 'System.Collections.Generic.IEnumerable`1<System.String> Microsoft.Extensions.Configuration.IConfigurationProvider.GetKeys(System.String)'.
   at Microsoft.Extensions.Configuration.InternalConfigurationRootExtensions.<>c__DisplayClass0_0.<GetChildrenImplementation>b__0(IConfigurationProvider p)
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext() in D:\runtime\src\libraries\System.Linq\src\System\Linq\SelectMany.cs:line 184
   at System.Collections.Generic.HashSet`1.UnionWith(IEnumerable`1 other) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\Collections\Generic\HashSet.cs:line 490
   at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\Collections\Generic\HashSet.cs:line 101
   at System.Linq.Enumerable.DistinctIterator`1.ToArray() in D:\runtime\src\libraries\System.Linq\src\System\Linq\Distinct.SpeedOpt.cs:line 12
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source) in D:\runtime\src\libraries\System.Linq\src\System\Linq\Buffer.cs:line 30
   at System.Linq.OrderedEnumerable`1.GetEnumerator()+MoveNext() in D:\runtime\src\libraries\System.Linq\src\System\Linq\OrderedEnumerable.cs:line 24
   at System.Linq.Enumerable.SelectIPartitionIterator`2.MoveNext() in D:\runtime\src\libraries\System.Linq\src\System\Linq\Select.SpeedOpt.cs:line 521
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source) in D:\runtime\src\libraries\System.Linq\src\System\Linq\AnyAll.cs:line 43
   at Microsoft.Extensions.Configuration.ConfigurationBinder.BindInstance(Type type, Object instance, IConfiguration config, BinderOptions options)
   at Microsoft.Extensions.Configuration.ConfigurationBinder.Get(IConfiguration configuration, Type type, Action`1 configureOptions)
   at Microsoft.Extensions.Configuration.ConfigurationBinder.Get[T](IConfiguration configuration, Action`1 configureOptions)
   at Microsoft.Extensions.Configuration.ConfigurationBinder.Get[T](IConfiguration configuration)
   at Microsoft.Extensions.Configuration.ConfigurationBinderBenchmarks.Get() in C:\performance\src\benchmarks\micro\libraries\Microsoft.Extensions.Configuration\ConfigurationBinderBenchmarks.cs:line 44

```

But I think it can be resolved by letting nuget package `Microsoft.Extensions.Configuration` requires a newer version of `Microsoft.Extensions.Configuration.Abstractions`?

Reply:
Change to default interface method.